File: solutions/go/bowling/1/bowling.go
package bowling

import "fmt"

const (
	pinsPerFrame  = 10
	framesPerGame = 10
	rollsPerFrame = 2
)

type Game struct {
	framesCompleted int
	score           int
	pinsUp          int
	rolls           int
	bonus1          int
	bonus2          int
}

func NewGame() *Game {
	return &Game{pinsUp: pinsPerFrame}
}

func (game *Game) isComplete() bool {
	return game.framesCompleted >= framesPerGame && game.bonus1 == 0 && game.bonus2 == 0
}

func (game *Game) Roll(pins int) error {
	if pins < 0 {
		return fmt.Errorf("can't roll negative pins %d", pins)
	}
	if pins > game.pinsUp {
		return fmt.Errorf("there are only %d pins up, can't roll %d", game.pinsUp, pins)
	}
	if game.isComplete() {
		return fmt.Errorf("game complete, can't roll %d", pins)
	}

	game.rolls++
	game.pinsUp -= pins
	multiplier := 1
	isFill := game.framesCompleted >= framesPerGame
	if isFill {
		multiplier = 0
	}
	multiplier += game.bonus1 + game.bonus2
	game.score += pins * multiplier
	game.bonus1 = game.bonus2
	game.bonus2 = 0

	if game.pinsUp == 0 || game.rolls == rollsPerFrame {
		if game.pinsUp == 0 && !isFill {
			if game.rolls == rollsPerFrame {
				game.bonus1++
			} else {
				game.bonus2++
			}
		}
		game.rolls = 0
		game.pinsUp = pinsPerFrame
		game.framesCompleted++
	}
	return nil
}

func (game *Game) Score() (int, error) {
	if !game.isComplete() {
		return 0, fmt.Errorf("game incomplete: On frame %d", game.framesCompleted+1)
	}
	return game.score, nil
}

File: solutions/python/bowling/1/bowling.py
class BowlingGame:
    def __init__(self):
        self.throws = []
        self.frame = 20
        self.bonus_frame = 0

    def roll(self, pins):
        if pins > 10 or pins < 0:
            raise ValueError("Pins should be inbetween 0 and 10")
        self.throws.append(pins)
        if self.bonus_frame:
            self.bonus_frame -= 1
            if all(
                (
                    self.bonus_frame == 0,
                    self.throws[-3] == 10,
                    self.throws[-2] < 10,
                    self.throws[-2] + self.throws[-1] > 10,
                )
            ):
                raise ValueError("There should only be 10 pins")
            return
        if pins == 10:
            self.frame -= 2
            if self.frame == 0:
                self.bonus_frame = 2
        else:
            self.frame -= 1
            if self.frame % 2 == 0 and self.throws[-2] + self.throws[-1] > 10:
                raise ValueError("There should only be 10 pins")
            if self.frame == 0 and self.throws[-2] + self.throws[-1] == 10:
                self.bonus_frame = 1
        if self.frame < 0:
            raise ValueError("Game has already ended")

    def score(self):
        if not 10 < len(self.throws) < 22 or self.bonus_frame != 0:
            raise ValueError("Invalid Game")
        point = 0
        for index, _ in enumerate(self.throws):
            if self.throws[index] == 10:
                self.frame += 2
                point += sum(self.throws[index : index + 3])
            else:
                self.frame += 1
                if (
                    self.frame % 2 == 0
                    and self.throws[index] + self.throws[index - 1] == 10
                ):
                    point += sum(self.throws[index : index + 2])
                else:
                    point += self.throws[index]
            if self.frame == 20:
                break
        return point
